File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/PeakFinding.java
package com.divideconquer.algorithms;

public final class PeakFinding {
    private PeakFinding() {
    }

    public static int peakIndex(int[] values) {
        if (values == null || values.length == 0) {
            throw new IllegalArgumentException("values must not be empty");
        }
        return find(values, 0, values.length - 1);
    }

    private static int find(int[] values, int left, int right) {
        if (left == right) {
            return left;
        }
        int mid = left + (right - left) / 2;
        if (values[mid] < values[mid + 1]) {
            return find(values, mid + 1, right);
        }
        return find(values, left, mid);
    }
}

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/peak_finding.py
"""One-dimensional peak finding."""


def peak_index(values: list[int]) -> int:
    """Return an index whose value is not smaller than its neighbors."""
    if not values:
        raise ValueError("values must not be empty")
    return _find(values, 0, len(values) - 1)


def _find(values: list[int], left: int, right: int) -> int:
    if left == right:
        return left
    mid = left + (right - left) // 2
    if values[mid] < values[mid + 1]:
        return _find(values, mid + 1, right)
    return _find(values, left, mid)

